test/storage/offline_database.cpp

```diff
@@ -554,23 +554,85 @@ static int databaseUserVersion(const std::string& path) {
     return stmt.get<int>(0);
 }
 
+static std::string databaseJournalMode(const std::string& path) {
+    mapbox::sqlite::Database db(path, mapbox::sqlite::ReadOnly);
+    mapbox::sqlite::Statement stmt = db.prepare("pragma journal_mode");
+    stmt.run();
+    return stmt.get<std::string>(0);
+}
+
+static int databaseSyncMode(const std::string& path) {
+    mapbox::sqlite::Database db(path, mapbox::sqlite::ReadOnly);
+    mapbox::sqlite::Statement stmt = db.prepare("pragma synchronous");
+    stmt.run();
+    return stmt.get<int>(0);
+}
+
 TEST(OfflineDatabase, MigrateFromV2Schema) {
     using namespace mbgl;
 
     // v2.db is a v2 database containing a single offline region with a small number of resources.
 
-    deleteFile("test/fixtures/offline_database/v4.db");
-    writeFile("test/fixtures/offline_database/v4.db", util::read_file("test/fixtures/offline_database/v2.db"));
+    deleteFile("test/fixtures/offline_database/v5.db");
+    writeFile("test/fixtures/offline_database/v5.db", util::read_file("test/fixtures/offline_database/v2.db"));
 
     {
-        OfflineDatabase db("test/fixtures/offline_database/v4.db", 0);
+        OfflineDatabase db("test/fixtures/offline_database/v5.db", 0);
         auto regions = db.listRegions();
         for (auto& region : regions) {
             db.deleteRegion(std::move(region));
         }
     }
 
-    EXPECT_EQ(4, databaseUserVersion("test/fixtures/offline_database/v4.db"));
-    EXPECT_LT(databasePageCount("test/fixtures/offline_database/v4.db"),
+    EXPECT_EQ(5, databaseUserVersion("test/fixtures/offline_database/v5.db"));
+    EXPECT_LT(databasePageCount("test/fixtures/offline_database/v5.db"),
               databasePageCount("test/fixtures/offline_database/v2.db"));
 }
+
+TEST(OfflineDatabase, MigrateFromV3Schema) {
+    using namespace mbgl;
+
+    // v3.db is a v3 database, migrated from v2.
+
+    deleteFile("test/fixtures/offline_database/v5.db");
+    writeFile("test/fixtures/offline_database/v5.db", util::read_file("test/fixtures/offline_database/v3.db"));
+
+    {
+        OfflineDatabase db("test/fixtures/offline_database/v5.db", 0);
+        auto regions = db.listRegions();
+        for (auto& region : regions) {
+            db.deleteRegion(std::move(region));
+        }
+    }
+
+    EXPECT_EQ(5, databaseUserVersion("test/fixtures/offline_database/v5.db"));
+    EXPECT_LT(databasePageCount("test/fixtures/offline_database/v5.db"),
```

Comment:
✂️ this assertion (and in `MigrateFromV4Schema` too). It's specific to the v2 ⇢ v3 migration (checking that it runs `VACUUM`). If it passes for other migrations, it's coincidental.